### ISSUE TITLE
Justification results in leading space

### ISSUE THREAD

**Issue by mattfbacon** [typst/typst#1794]
### Description

Here is an MRE:

```typst
#set par(justify: true)
#set text(size: 12pt)
#set page(width: 45mm, height: auto)

lorem ipsum 1234, lorem ipsum dolor sit amet
```

![MRE](https://github.com/typst/typst/assets/58113890/d3a6c5a3-1294-4cb1-b58d-1297624cfc0d)

This was built using the latest typst CLI fresh from git.

I have found that this occurs when a line is justified after breaking on a number followed by a comma or period.

### Operating system

Linux

### Typst version

- [X] I am using the latest version of Typst

**Comment by bluebear94** [comment]
Should leading spaces show up at all? That is, should the following two lines produce different outputs:

```
#"abc"
#"  abc"
```

**Comment by laurmaedje** [comment]
It's also frustrating if you can't force any whitespace, so I think it should. It's already a problem that _trailing_ whitespace is always trimmed: #597

**Comment by mattfbacon** [comment]
> two consecutive breakpoints

I think I understand.. but how is the number related? I could't reproduce without it.

**Comment by zouharvi** [comment]
I am not sure if this is the same problem but I get this sometimes when the line ends with a citation:

<img width="350em" src="https://github.com/typst/typst/assets/7661193/a030924d-752d-4d02-9a43-3684d74d5051">

**Comment by mattfbacon** [comment]
Here's a full list of characters that produce such a breakpoint:

'\n', '\u{b}', '\u{c}', '\r', ')', ',', '.', '/', ':', ';', ']', '}', '\u{85}', ';', '։', '،', '؍', '߸', '༻', '༽', '᚜', '\u{2028}', '\u{2029}', '⁄', '⁆', '⁾', '₎', '⌉', '⌋', '〉', '❩', '❫', '❭', '❯', '❱', '❳', '❵', '⟆', '⟧', '⟩', '⟫', '⟭', '⟯', '⦄', '⦆', '⦈', '⦊', '⦌', '⦎', '⦐', '⦒', '⦔', '⦖', '⦘', '⧙', '⧛', '⧽', '⸣', '⸥', '⸧', '⸩', '⹖', '⹘', '⹚', '⹜', '、', '。', '〉', '》', '」', '』', '】', '〕', '〗', '〙', '〛', '〞', '〟', '﴾', '︐', '︑', '︒', '︓', '︔', '︘', '︶', '︸', '︺', '︼', '︾', '﹀', '﹂', '﹄', '﹈', '﹐', '﹒', '﹚', '﹜', '﹞', '）', '，', '．', '］', '｝', '｠', '｡', '｣', '､', '𓉛', '𓉜', '𓉝', '𓊂', '𓊇', '𓊉', '𓍺', '𓍻', '\u{13438}', '\u{1343d}', '\u{1343f}', '𔗏'

**Comment by raphCode** [comment]
I am still seeing this bug in typst 0.8.0 (360cc9b9).
Based on the dates where the PR was merged / the upstream bug in icu4x fixed, I would have expected that the issue is fixed in 0.8.0?

**Comment by laurmaedje** [comment]
So, #1821 was only half of the fix (for justification), but #2056 should also have fixed this for non-justified text. Both PRs are workarounds around the upstream icu4x issue, which has been fixed in the meantime, but we have not yet regenerated the icu4x data files. But both PRs also happened before 0.8.0, so the issue should not have appeared in 0.8.0. 

@raphCode Do you have a reproduction for the issue on 0.8.0?
